Spark.NET: new overview docs (#14755)
dotnet/docs: docs/spark/what-is-apache-spark-dotnet.md

description: Learn about .NET for Apache Spark, a free, open-source, and cross-platform big data analytics framework that takes Spark anywhere you write .NET code.
author: mamccrea
ms.topic: overview
ms.date: 09/27/2019
---

# What is .NET for Apache Spark?

> [Apache Spark](what-is-spark.md) is a general-purpose distributed processing engine for analytics over large data sets - typically terabytes or petabytes of data. With .NET for Apache Spark, a free, open-source, and cross-platform big data analytics framework, you can now add the power of Apache Spark to your big data applications using languages you already know.

Reviewer: I am torn here. .NET for Apache Spark is not the analytics framework per se (that is Apache Spark) but allows .NET users to use it. What about:

With .NET for Apache Spark, the free, open-source, and cross-platform .NET Support for the popular open-source big data analytics framework, you can now add the power of Apache Spark to your big data applications using languages you already know.

## Why choose .NET for Apache Spark?

> .NET for Apache Spark provides high performance APIs for using Spark from C# and F#. With C# and F#, you can access:

Reviewer 2: The goal of .NET Spark is empowering developers with .NET knowledge/code bases to participate in the world of big data, and a non-goal is to convert devs with expertise in other languages to choose .NET instead. Thus, I think we'd want to emphasize this in this section titled "Why choose .NET for Apache Spark?" so that it doesn't come across that we're trying to convert devs from other languages. 

Maybe rephrasing this opening sentence/introducing a new sentence like:
".NET for Apache Spark empowers developers with .NET experience or code bases to participate in the world of big data analytics. .NET for Apache Spark provides high performance APIs..."

> * DataFrame and SparkSQL for working with structured data

Reviewer 2: For all of the scenarios in this list (structured/batch data, streaming, SQL, ML), it would be beneficial to provide some more explanation/context- when I was new to big data, I didn't know what any of this was, and it'd be great to explain it here so that users don't leave confused or have to search on the Internet for further definitions.

Maybe sub-bullets, an additional section, or just longer bullets defining these scenarios/data types: batch data, structured data, Spark SQL, ML training and prediction

* Spark Structured Streaming for working with streaming data

> * Spark SQL

Reviewer 2: We'd probably want to more clearly differentiate Spark SQL (writing queries with SQL syntax) from SparkSQL (for working with structured data)

> * MLlib

Reviewer 2: We don't currently support MLlib (although it may be on the road map). We might want to rephrase this as just machine learning, and could mention ML.NET integration:

```suggestion
* Machine learning integration for faster training and prediction (i.e. use .NET for Apache Spark alongside [ML.NET](http://dot.net/ml))
```

> .NET for Apache Spark is compliant with .NET Standard, a formal specification of .NET APIs that are common across .NET implementations. This means you can use .NET for Apache Spark anywhere you write .NET code allowing you to reuse all the knowledge, skills, code, and libraries you already have as a .NET developer.

Reviewer 2: ```suggestion
.NET for Apache Spark is compliant with .NET Standard, a formal specification of .NET APIs that are common across .NET implementations. This means you can use .NET for Apache Spark anywhere you write .NET code, allowing you to reuse all the knowledge, skills, code, and libraries you already have as a .NET developer.
```

.NET for Apache Spark runs on Windows, Linux, and macOS using .NET Core. It also runs on Windows using .NET Framework. You can deploy your applications to all major cloud providers including Azure HDInsight Spark, Amazon EMR Spark, Azure Databricks, and Databricks on AWS.

## .NET for Apache Spark architecture

> The C#/ F# language binding to Spark is written on a new Spark interop layer which offers easier extensibility. This new layer of Spark interop was written using the best practices for language extension and optimizes for interop and performance. Long term this extensibility can be used for adding support for other languages in Spark.

Reviewer 3: ```suggestion
The C#/ F# language binding to Spark is written on a new Spark interop layer which offers easier extensibility. This new layer of Spark interop was written using the best practices for language extension and optimizes for interop and performance. Long term, this extensibility can be used for adding support for other languages in Spark.
```

> [!div class="mx-imgBorder"]
> ![.NET for Apache Spark architecture](media/dotnet-spark-architecture.png)

> You can about interop support for Spark language extensions from [the proposal](https://issues.apache.org/jira/browse/SPARK-26257).

Reviewer 3: You can learn?

## .NET for Apache Spark performance

> When compared against Python and Scala using [TPC-H benchmark](http://www.tpc.org/tpch/), .NET for Apache Spark performs well in most cases and is 2x faster than Python when user-defined function performance is critical. There is an ongoing effort to improve and benchmark performance. 

Reviewer: ```suggestion
When compared against Python and Scala using the [TPC-H benchmark](http://www.tpc.org/tpch/), .NET for Apache Spark performs well in most cases: if no user-defined functions are present, it performs as fast as Scala and Python, and it is 2x faster than Python when user-defined function performance is critical. The .NET for Apache Spark project is continuing to improve and benchmark performance. 
```

To do your own benchmarking, see the benchmarks available on the [.NET for Apache Spark GitHub](https://github.com/dotnet/spark/tree/master/benchmark).

## .NET for Apache Spark roadmap

Learn about short term and long term plans from the official [.NET for Apache Spark roadmap](https://github.com/dotnet/spark/blob/master/ROADMAP.md).

## .NET Foundation

The .NET for Apache Spark project is part of the [.NET Foundation](https://www.dotnetfoundation.org/).

## Contributions

The .NET for Apache Spark team encourages contributions, both GitHub issues and pull requests. First, look for an [existing issue](https://github.com/dotnet/spark/issues). If you can't find an existing issue, [open a new issue](https://github.com/dotnet/spark/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+).

## Next steps

Try .NET for Apache Spark.
> [!div class="nextstepaction"]

> > [Tutorial: Get started with .NET for Apache Spark](../tutorials/get-started.md)

Reviewer 3: ```suggestion
> [Tutorial: Get started with .NET for Apache Spark](./tutorials/get-started.md)
```
